### ISSUE TITLE
java.net.UnknownHostException due to AWS EC2 instance hostname not in /etc/hosts

### ISSUE THREAD

**Issue by dumkydewilde** [snowplow-devops/terraform-aws-enrich-kinesis-ec2#30]
I recently set up the secure version of the AWS Quickstart terraform and upgraded all the Terraform modules to their latest version. The only issue I kept running is that the enricher seems to be calling its own `$HOSTNAME` at some point (e.g. `ip-10-10-3-119` or `ip-10-10-3-119.us-west-2.compute.internal`) however that hostname is not available in `/etc/hosts` so it throws and UnknownHostException. This is described in more detail on [this Discourse thread](https://discourse.snowplow.io/t/unknown-host-exception-in-snowplow-open-source-quick-start/6979).

I have solved this by forking this Terraform module and adding a `custom_script` variable like so:

```
# main.tf

module "enrich_kinesis" {
  source  = "./enrich-kinesis-ec2-aws"
  #version = "0.5.1"

  ...

  custom_script = "echo \"127.0.0.1  $HOSTNAME\" >> /etc/hosts"
}
```

This is then referenced `user-data.sh.tmpl` to make sure it runs when starting the EC2 instances:

```sh
# user-data.sh.tmpl
readonly CONFIG_DIR=/opt/snowplow/config

sudo mkdir -p $${CONFIG_DIR}
sudo cat << EOF > $${CONFIG_DIR}/enrich.hocon
${config}
EOF

sudo docker run \
... 

${telemetry_script}

${custom_script}
```

This works for now, but I have a few considerations (and am happy to contribute where possible/needed):
- Is there an underlying reason why the instance hostname is referenced in this way?
- Is there an underlying reason why the instance hostname is not available to the Java runtime/docker container?
- Might it be generally convenient to be able to add a custom script to the startup script of all EC2 instances just like it's possible to turn telemetry on/off
- I've only run into this for the enricher, not for any other services using EC2

One more observation, not sure if it's relevant is that I haven't been able to use the method in [/snowplow-devops/terraform-aws-service-ec2/](https://github.com/snowplow-devops/terraform-aws-service-ec2/blob/main/templates/user-data.sh.tmpl) to get the instance ID:
```
function get_instance_id() {
  local __ec2_metadata_token
  __ec2_metadata_token=$(curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600")
  curl -H "X-aws-ec2-metadata-token: $${__ec2_metadata_token}" --silent --location "http://169.254.169.254/latest/meta-data/instance-id/"
}
```

However, I was able to use either `$HOSTNAME` (for just the hostname) or the instance's local ec2-metadata utility `/opt/aws/bin/ec2-metadata -i` to get the instance ID.

**Comment by dumkydewilde** [comment]
As an FYI, I'm seeing the same problem in the [EC2 Snowflake Loader](https://github.com/snowplow-devops/terraform-aws-snowflake-loader-ec2).

**Comment by jbeemster** [comment]
Hey @dumkydewilde having tested these modules extensively I have not run into these issues.  Could you share how your VPC is configured?  Can you confirm this only happens with your custom VPC and not with the default VPC that comes with the sub-account?

**Comment by dumkydewilde** [comment]
Thanks so much @jbeemster! I had indeed not run into this previously, but I was using older versions. Your comment indeed made me go back to my VPC config and investigate further because in this case I wasn't able to set up the VPC with Terraform due to some company restrictions. Turns out, the VPC was missing a DHCP option to use the internal domain names in the private subnet (e.g. that `ip-10-10-3-119.us-west-2.compute.internal`). I've since added this and it's working like a charm! 👌. Sorry to bother you with this, as it was indeed VPC and not Snowplow related, but let's hope this comment will save someone a few hours of digging in the future.
